src/compiler/types/lines_6000_8000.rs

```diff
@@ -231,4 +829,106 @@ pub enum EmitHint {
     Unspecified,
 }
 
-pub struct NodeFactory {}
+bitflags! {
+    pub struct OuterExpressionKinds: u32 {
+        const None = 0;
+        const Parentheses = 1 << 0;
+        const TypeAssertions = 1 << 1;
+        const NonNullAssertions = 1 << 2;
+        const PartiallyEmittedExpressions = 1 << 3;
+
+        const Assertions = Self::TypeAssertions.bits | Self::NonNullAssertions.bits;
+        const All = Self::Parentheses.bits | Self::Assertions.bits | Self::PartiallyEmittedExpressions.bits;
+
+        const ExcludeJSDocTypeAssertion = 1 << 4;
+    }
+}
+
+pub trait ParenthesizerRules {
+    // fn get_parenthesize_left_side_of_binary_for_operator(&self, binary_operator: SyntaxKind) ->
```

Comment:
These methods returning closures seemed like they'd be tricky to implement on a trait like this, so for now planning on just "inlining" this optimization (I guess always just calling `parenthesize_left_side_of_binary()`?)